### PR TITLE
Progress for Next.js 

### LINKED ISSUE #485: `bun create` to be compatible with `npm create`
The feature that I'm proposing is the support for `bun create` to create project using NPM packages.

For example, with npm this `npm create vite@latest` can be used to create a project using the latest version of vite, while bun doesn't support npm for his `create` command.

I realize that this project is still in its infancy, but I wanted to say to every developer of the project that your work is truly amazing.

### LINKED ISSUE #3620: Implement `child_process.fork()`
Spawns another process of Bun with IPC.

https://nodejs.org/api/child_process.html#child_processforkmodulepath-args-options

### LINKED ISSUE #4166: `bun run dev` acts differently than `bun dev`
### What version of Bun is running?

0.7.3

### What platform is your computer?

Linux 5.15.90.4-microsoft-standard-WSL2 x86_64 x86_64

### What steps can reproduce the bug?

1. `bun create react app`
2. `cd app`
3. `bun dev` => all static assets will be 404
4. `bun run dev` => works fine

### What is the expected behavior?

No 404s.

### What do you see instead?

404 for every static asset for `public` and `build` dir.

### Additional information

I tried to add some logs to `dev.tsx` and there is nothing logged when using `bun dev` command. It's like it's not even using that `dev.tsx` file.

### PR DESCRIPTION
**Pull request by paperclover**
### What does this PR do?

- [x] feat(runtime): implement `child_process.fork` / ipc channels. Closes #3620
- [x] fix(runtime): undefined values in spawn env args no longer be passed
- [x] chore: remove `BUN_OVERRIDE_MODULE_PATH` (it doesn't do anything)
- [x] feat(runtime): implement `fs.promises.opendir`
- [x] feat(runtime): implement `v8.getHeapStatistics`
- [X] fix(runtime): do not append `node:` to `require.resolve("without_prefix")`
- [X] feat(runtime): implement setter for `require('module')._resolveFilename`
- [X] fix(runtime): make `fs.WriteStream` and `NativeWritable` function-style classes 
- [X] fix(runtime): do not inline resolved paths into transpiled CommonJS (fixes next.js react duplication bugs)
- [X] fix(runtime): ensure `req.socket` === the `socket` passed to the `"upgrade"` event.
- [X] chore(cli): remove `bun dev`. closes #4166
- [X] feat(cli): make `bun create` execute `create-` packages. closes #485
- [X] chore(cli): make the cli help menus nicer.
- ... any more?

### How did you verify your code works?

I've been testing with Next.js to get it farther.

### CODE REVIEW

**Comment by codehz** [comment]
https://github.com/oven-sh/bun/issues/4236
will it expose same internal symbol (`Symbol(kChannelHandle)`) as node does?

**Comment by paperclover** [comment]
> #4236
> will it expose same internal symbol (`Symbol(kChannelHandle)`) as node does?

According to the [node docs on IPC](https://nodejs.org/dist/latest-v20.x/docs/api/child_process.html#optionsstdio):

> Accessing the IPC channel fd in any way other than [process.send()](https://nodejs.org/dist/latest-v20.x/docs/api/process.html#processsendmessage-sendhandle-options-callback) or using the IPC channel with a child process that is not a Node.js instance is not supported.

The internal implementation I'm working with so far does not look anything like node's at all, so even if this symbol existed with some value, it simply wouldn't work. All of the IPC code is done in native.

This library should use `"pipe"` with a fourth file descriptor (scroll down in the docs link) instead of hijacking the Node IPC system for messages, it would be much cleaner on their end... however Bun/this per doesn't implement extra fds yet. The only way it would work right now is using stdout+stdin pipes.

File: src/js/node/v8.ts [hunk 1]
```diff
@@ -32,7 +32,22 @@ function getHeapSnapshot() {
...
+    heap_size_limit: 4345298944,
+    malloced_memory: 262320,
+    peak_malloced_memory: 571392,
+    does_zap_garbage: 0,
+    number_of_native_contexts: 2,
+    number_of_detached_contexts: 0,
+    total_global_handles_size: 8192,
+    used_global_handles_size: 3488,
+    external_memory: 2136251,
+  };
```
Comment (paperclover) [inline src/js/node/v8.ts:52]: todo:

File: src/js/node/fs.promises.ts [hunk 2]
```diff
@@ -105,6 +105,19 @@ function cp(src, dest, options) {
   return fs.cp(src, dest, options.recursive, options.errorOnExist, options.force ?? true, options.mode);
 }
 
+function opendir(dir: string) {
+  // TODO: actually implement this 💀
```
Comment (paperclover) [inline src/js/node/fs.promises.ts]: todo:

**Comment by birkskyum** [comment]
Closes https://github.com/oven-sh/bun/issues/3620

File: test/js/node/fs/fs.test.ts [hunk 3]
```diff
@@ -2025,7 +2025,7 @@ it("BigIntStats", () => {
 it("test syscall errno, issue#4198", () => {
   const path = `${tmpdir()}/non-existent-${Date.now()}.txt`;
   expect(() => openSync(path, "r")).toThrow("No such file or directory");
-  expect(() => readSync(2147483640, Buffer.alloc(0))).toThrow("Bad file number");
+  expect(() => readSync(2147483640, Buffer.alloc(0))).toThrow("EBADF");
```
Comment (Hanaasagi) [inline test/js/node/fs/fs.test.ts]:
```suggestion
  if (os.platform() == "darwin") {
    expect(() => readSync(2147483640, Buffer.alloc(0))).toThrow("Bad file descriptor");
  } else {
    expect(() => readSync(2147483640, Buffer.alloc(0))).toThrow("Bad file number");
  }

```

`os` is not imported.

**Comment by birkskyum** [comment]
Tried to test this canary release, but it seems like something broke regarding macos arm64. https://github.com/oven-sh/bun/actions/runs/6109378228/job/16580283046

**Comment by Grandnainconnu** [comment]
`process.send` still not working with Bun `1.0.6` on macos arm64

**Comment by paperclover** [comment]
process.send is only defined when IPC is enabled. make sure you are using child_processs.fork to enable it, and note it only works for bun <-> bun communication right now.

**Pull request merged**